src/groupeddataframe/grouping.jl

```diff
@@ -115,96 +114,204 @@ Base.getindex(gd::GroupedDataFrame, I::AbstractArray{Bool}) =
-end
-
 wrap(df::AbstractDataFrame) = df
+wrap(nt::NamedTuple) = nt
 wrap(A::Matrix) = convert(DataFrame, A)
-wrap(s::Any) = DataFrame(x1 = s)
+wrap(s::Union{AbstractVector, Tuple}) = DataFrame(x1 = s)
+wrap(s::Any) = (x1 = s,)
 
 """
-Combine a GroupApplied object (rudimentary)
+Apply a function to each group of rows and combine the result
 
 ```julia
-combine(ga::GroupApplied)
+map(f::Function, gd::GroupedDataFrame)
 ```
 
 ### Arguments
 
-* `ga` : a GroupApplied
+* `gd` : a `GroupedDataFrame` object
 
 ### Returns
 
 * `::DataFrame`
 
+### Details
+
+For each group in `gd`, `f` is passed a `SubDataFrame` view with the corresponding rows.
+`f` can return a single value, a named tuple, a vector, or a data frame.
+This determines the shape of the resulting data frame:
+- A single value gives a data frame with a single column and one row per group.
+- A named tuple gives a data frame with one column for each field in the named tuple
+  and one row per group.
+- A vector gives a data frame with a single column and as many rows
+  for each group as the length of the returned vector for that group.
+- A data frame gives a data frame with the same columns and as many rows
+  for each group as the rows returned for that group.
+
+In all cases, the resulting data frame contains all the grouping columns in addition
+to those listed above. Note that `f` must always return the same type of object for
+all groups, and (if a named tuple or data frame) with the same fields or columns in the
+same order. Returning a single value or a named tuple is significantly faster than
+returning a vector or a data frame.
+
+The resulting data frame will be sorted if `sort=true` was passed to the [`groupby`](@ref)
+call from which `gd` was constructed. Otherwise, ordering of rows is undefined.
+
 ### Examples
 
 ```julia
 df = DataFrame(a = repeat([1, 2, 3, 4], outer=[2]),
                b = repeat([2, 1], outer=[4]),
                c = randn(8))
 gd = groupby(df, :a)
-combine(map(d -> mean(skipmissing(d[:c])), gd))
+map(d -> sum(skipmissing(d[:c])), gd)
 ```
 
+### See also
+
+`by(f, df, cols)` is a shorthand for `map(f, groupby(df, cols))`.
+
 """
-function combine(ga::GroupApplied)
-    gd, vals = ga.gd, ga.vals
-    valscat = _vcat(vals)
-    idx = Vector{Int}(undef, size(valscat, 1))
-    j = 0
-    @inbounds for (start, val) in zip(gd.starts, vals)
-        n = size(val, 1)
-        idx[j .+ (1:n)] .= gd.idx[start]
-        j += n
+function Base.map(f::Function, gd::GroupedDataFrame)
+    if length(gd) > 0
+        idx, valscat = _combine(wrap(f(gd[1])), f, gd)
+        return hcat!(gd.parent[idx, gd.cols], valscat)
+    else
+        return similar(gd.parent, 0)[gd.cols]
     end
-    hcat!(gd.parent[idx, gd.cols], valscat)
 end
 
+function _combine(first::NamedTuple, f::Function, gd::GroupedDataFrame)
+    m = length(first)
+    n = length(gd)
+    idx = Vector{Int}(undef, n)
+    initialcols = ntuple(i -> Vector{typeof(first[i])}(undef, n), m)
```

Comment:
What about CategoricalValues here? Don't we want to use some kind of `similar` construct? In Tables.jl, we have [`allocatecolumn`](https://github.com/JuliaData/Tables.jl/blob/master/src/fallbacks.jl#L47) which can be overloaded by custom scalars to provide an alternative AbstractVector type. It'd be nice possibly have something like `vectortype(scalar::T) = Vector{T}` in Base eventually so this could be used throughout the package ecosystem.

Reply:
Right. It's ironic I missed that point. :-)

I've added a commit to use `similar` for the data frame case, but as you note it can't be solved for the scalar case without calling `allocatecolumn`. I guess we can wait until your Tables.jl PR is merged before merging this one. Do you think it would be OK for Tables.jl to implement the necessary method? It could use Require.jl to avoid a hard dependency on CategoricalArrays.

I also agree it would be useful to have something like this in Base.

Reply:
Yeah, Tables.jl already defines the [right methods](https://github.com/JuliaData/Tables.jl/blob/master/src/Tables.jl#L11) for both CategoricalValue & WeakRefString.

Reply:
Perfect. So I can update the PR once you've merged yours.

Reply:
Done. That allowed simplifying constructors a bit too.